drivers/bt_ll_nrfxlib/flash/flash_ll_nrfxlib.c

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <string.h>
+#include <errno.h>
+#include <zephyr.h>
+#include <device.h>
+#include <soc.h>
+#include <flash.h>
+#include <misc/util.h>
+
+#include <nrfx_nvmc.h>
+
+#include "ble_controller_soc.h"
+#include "multithreading_lock.h"
+
+/* NOTE: The driver supports unligned writes, but some file systems (like FCB)
+ * may use the driver sub-optimally as a result. Word aligned writes are faster
+ * and require less overhead. This value can be changed to 4 to minimize this
+ * overhead.
+ */
+#define FLASH_DRIVER_WRITE_BLOCK_SIZE 1
+
+static struct {
+	/** Used to ensure a single ongoing operation at any time.  */
+	struct k_mutex lock;
+	/** Used to wait for the asynchronous operation to complete */
+	struct k_sem sync;
+	const void *data;
+	off_t addr;
+	u16_t len;
+	u16_t prev_len;
+	u32_t tmp_word;      /**< Used for unalinged writes. */
+	/* NOTE: Read is not async, so not a part of this enum. */
+	enum {
+		FLASH_OP_NONE,
+		FLASH_OP_WRITE,
+		FLASH_OP_ERASE
+	} op;
+} flash_state;
+
+/* Forward declarations */
+static int btctlr_flash_read(struct device *dev,
+			     off_t offset,
+			     void *data,
+			     size_t len);
+static int btctlr_flash_write(struct device *dev,
+			      off_t offset,
+			      const void *data,
+			      size_t len);
+static int btctlr_flash_erase(struct device *dev,
+			      off_t offset,
+			      size_t size);
+static int btctlr_flash_write_protection_set(struct device *dev,
+					     bool enable);
+
+static int flash_op_execute(void);
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void btctlr_flash_page_layout_get(
+	struct device *dev,
+	const struct flash_pages_layout **layout,
+	size_t *layout_size);
+#endif /* defined(CONFIG_FLASH_PAGE_LAYOUT) */
+
+static const struct flash_driver_api btctrl_flash_api = {
+	.read = btctlr_flash_read,
+	.write = btctlr_flash_write,
+	.erase = btctlr_flash_erase,
+	.write_protection = btctlr_flash_write_protection_set,
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = btctlr_flash_page_layout_get,
+#endif  /* CONFIG_FLASH_PAGE_LAYOUT */
+	.write_block_size = FLASH_DRIVER_WRITE_BLOCK_SIZE
+};
+
+
+/* Utility functions */
+static inline bool is_addr_valid(off_t addr, size_t len)
+{
+	if ((addr < 0) ||
+	    (len > nrfx_nvmc_flash_size_get()) ||
+	    (addr >  nrfx_nvmc_flash_size_get() - len)) {
+		return false;
+	}
+
+	return true;
+}
+
+static inline bool is_aligned_32(off_t addr)
+{
+	return ((addr & 0x3) == 0);
+}
+
+static inline off_t align_32(off_t addr)
+{
+	return (addr & ~0x3);
+}
+
+static inline size_t bytes_to_words(size_t bytes)
+{
+	return bytes / sizeof(u32_t);
+}
+
+static inline bool is_page_aligned(off_t addr)
+{
+	return (addr & (nrfx_nvmc_flash_page_size_get() - 1)) == 0;
+}
+
+static void flash_operation_complete_callback(u32_t status)
```

Comment:
This is called from ble_controller. Can you tell from which context? Is is interrupt context?
If so it can happen that `flash_op_execute` that is called from this function will not be able to take the lock as it would still be taken (if timing is right).

Reply:
It's called from the low priority process interrupt used by the controller:
https://github.com/thomasstenersen/fw-nrfconnect-nrf/blob/ble-controller-0.2.0/subsys/bluetooth/controller/hci_driver.c#L342
https://github.com/thomasstenersen/fw-nrfconnect-nrf/blob/ble-controller-0.2.0/subsys/bluetooth/controller/hci_driver.c#L437
https://github.com/thomasstenersen/fw-nrfconnect-nrf/blob/ble-controller-0.2.0/subsys/bluetooth/controller/hci_driver.c#L22

Reply:
thx